support-frontend/app/services/PayPalNvpService.scala

```diff
@@ -90,14 +92,48 @@ class PayPalNvpService(apiConfig: PayPalConfig, wsClient: WSClient) extends Touc
       "RETURNURL" -> returnUrl,
       "CANCELURL" -> cancelUrl,
       "BILLINGTYPE" -> "MerchantInitiatedBilling",
-      "NOSHIPPING" -> "1"
+      "NOSHIPPING" -> noShipping
     )
 
     nvpRequest(paymentParams).map { resp =>
       retrieveNVPParam(resp, "TOKEN")
     }
   }
 
+  // Sets up a payment by contacting PayPal and also fetches the users details.
+  def createAgreementAndRetrieveUser(token: Token): Future[Option[PayPalCheckoutDetails]] =
+    createBillingAgreement(token).flatMap(
+      maybeBaid =>
+      retrieveUserInformation(token).map(maybeUserDetails =>
+        maybeBaid match {
+          case Some(baid) => Some(PayPalCheckoutDetails(baid, maybeUserDetails))
+          case _ => None
+        }
+      )
+    )
+
+  def retrieveUserInformation(token: Token): Future[Option[PayPalUserDetails]] = {
+    val paymentParams = Map(
+      "METHOD" -> "GetExpressCheckoutDetails",
+      "TOKEN" -> token.token
+    )
+
+    nvpRequest(paymentParams).map { resp =>
+      for {
+        firstName <- retrieveNVPParam(resp, "FIRSTNAME")
+        lastName <- retrieveNVPParam(resp, "LASTNAME")
+        email <- retrieveNVPParam(resp, "EMAIL")
+        shipToStreet <- retrieveNVPParam(resp, "PAYMENTREQUEST_0_SHIPTOSTREET")
+        shipToCity <- retrieveNVPParam(resp, "PAYMENTREQUEST_0_SHIPTOCITY")
+        shipToState <- retrieveNVPParam(resp, "PAYMENTREQUEST_0_SHIPTOSTATE")
+        shipToZip <- retrieveNVPParam(resp, "PAYMENTREQUEST_0_SHIPTOZIP")
+        shipToCountryCode <- retrieveNVPParam(resp, "PAYMENTREQUEST_0_SHIPTOCOUNTRYCODE")
+      } yield PayPalUserDetails(
```

Comment:
if any are missing then we will get a None. is there any logging that's going to let us know what's going on, before we return a BadRequest?

Reply:
Good point I've added some logging